### ISSUE TITLE
Wrong Android resources found - App works with rebuild, doesn't work with build

### ISSUE THREAD

**Issue by stmax82** [xamarin/xamarin-android#1585]
We're haunted by a weird problem in one of our apps and I think I was finally able to reduce the project to a small post-able example.

A layout in an Android library defines a SurfaceView with the id "cameraPreview":

```xml
    <SurfaceView
        android:id="@+id/cameraPreview"
        android:layout_width="match_parent"
        android:layout_height="match_parent" />
```

Then the problem happens with the following line of code:

```csharp
var preview = FindViewById<SurfaceView>(Resource.Id.cameraPreview);
```

* Sometimes this line correctly returns a SurfaceView instance.
* Sometimes it returns null (and causes NullReferenceExceptions further down).
* Sometimes it throws an InvalidCastException because it tries to return an instance of something else.

We are not sure what causes the different behaviours. There seems to be a difference between Build and Rebuild and Clean+Build before starting the app. But we also have a project where the problem always happens, no matter if we Build, Rebuild or Clean+Build...

### Steps to Reproduce

* Please download (or git clone) the following example: https://github.com/stmax82/XamarinBuildVsRebuild.git
* Connect an Android phone to the PC
* Open MyApp.sln
* Set as StartUp Project "MyApp.Android"
* Make sure the phone is selected (normally happens automatically)
* Press F5 to start the app on the phone
* Once started, press the "TEST" button

**-> It will either fail with a NullReferenceException or with an InvalidCastException**

Now **delete everything** (this is important) and start over:

* Please download (or git clone) the following example: https://github.com/stmax82/XamarinBuildVsRebuild.git
* Connect an Android phone to the PC
* Open MyApp.sln
* Set as StartUp Project "MyApp.Android"
* Make sure the phone is selected (normally happens automatically)
* **Click "Clean Solution" on the solution**
* Press F5 to start the app on the phone
* Once started, press the "TEST" button

**-> This time it will work and display "Result: foo"**

### Expected Behavior

1. FindViewById should always return the SurfaceView with the id "cameraPreview"
2. Build vs Rebuild vs Clean+Build should always produce the exact same output

### Actual Behavior

* FindViewById sometimes returns wrong views or nothing (null) even though the view's id is defined
* Build produces a different output than Clean+Build and Rebuild

### Version Information

Microsoft Visual Studio Professional 2017 
Version 15.6.6
VisualStudio.15.Release/15.6.6+27428.2037
Microsoft .NET Framework
Version 4.7.02558

Installed Version: Professional

Visual F# Tools 10.1 for F# 4.1   00369-60000-00001-AA170
Microsoft Visual F# Tools 10.1 for F# 4.1

Application Insights Tools for Visual Studio Package   8.11.10402.2
Application Insights Tools for Visual Studio

ASP.NET and Web Tools 2017   15.0.40314.0
ASP.NET and Web Tools 2017

ASP.NET Core Razor Language Services   1.0
Provides languages services for ASP.NET Core Razor.

ASP.NET Web Frameworks and Tools 2017   5.2.51214.0
For additional information, visit https://www.asp.net/

Azure App Service Tools v3.0.0   15.0.40215.0
Azure App Service Tools v3.0.0

Azure Data Lake Node   1.0
This package contains the Data Lake integration nodes for Server Explorer.

Azure Data Lake Tools for Visual Studio   2.2.9000.1
Microsoft Azure Data Lake Tools for Visual Studio

Azure Data Lake Tools for Visual Studio   2.2.9000.1
Microsoft Azure Data Lake Tools for Visual Studio

Azure Functions and Web Jobs Tools   15.0.40215.0
Azure Functions and Web Jobs Tools

C# Tools   2.7.0-beta3-62715-05. Commit Hash: db02128e6e3c4bdfc93e6ec425ac9162b4d4fe80
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

Cookiecutter   15.6.18072.2
Provides tools for finding, instantiating and customizing templates in cookiecutter format.

Fabric.DiagnosticEvents   1.0
Fabric Diagnostic Events

JavaScript Language Service   2.0
JavaScript Language Service

Merq   1.1.17-rc (cba4571)
Command Bus, Event Stream and Async Manager for Visual Studio extensions.

Microsoft Azure HDInsight Azure Node   2.2.9000.1
HDInsight Node under Azure Node

Microsoft Azure Hive Query Language Service   2.2.9000.1
Language service for Hive query

Microsoft Azure Service Fabric Tools for Visual Studio   2.0
Microsoft Azure Service Fabric Tools for Visual Studio

Microsoft Azure Stream Analytics Language Service   2.2.9000.1
Language service for Azure Stream Analytics

Microsoft Azure Stream Analytics Node   1.0
Azure Stream Analytics Node under Azure Node

Microsoft Azure Tools   2.9
Microsoft Azure Tools for Microsoft Visual Studio 2017 - v2.9.51212.2

Microsoft Continuous Delivery Tools for Visual Studio   0.3
Simplifying the configuration of continuous build integration and continuous build delivery from within the Visual Studio IDE.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Microsoft MI-Based Debugger   1.0
Provides support for connecting Visual Studio to MI compatible debuggers

Microsoft Visual Studio Tools for Containers   1.1
Develop, run, validate your ASP.NET Core applications in the target environment. F5 your application directly into a container with debugging, or CTRL + F5 to edit & refresh your app without having to rebuild the container.

Mono Debugging for Visual Studio   4.9.11-pre (71eb098)
Support for debugging Mono processes with Visual Studio.

NuGet Package Manager   4.6.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit http://docs.nuget.org/.

ProjectServicesPackage Extension   1.0
ProjectServicesPackage Visual Studio Extension Detailed Info

Python   15.6.18072.2
Provides IntelliSense, projects, templates, debugging, interactive windows, and other support for Python developers.

Python - Django support   15.6.18072.2
Provides templates and integration for the Django web framework.

Python - IronPython support   15.6.18072.2
Provides templates and integration for IronPython-based projects.

Python - Profiling support   15.6.18072.2
Profiling support for Python projects.

R Tools for Visual Studio   1.3.40104.1351
Provides project system, R Interactive window, plotting, and more for the R programming language.

SQL Server Data Tools   15.1.61801.210
Microsoft SQL Server Data Tools

ToolWindowHostedEditor   1.0
Hosting json editor into a tool window

TypeScript Tools   15.6.20202.3
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   2.7.0-beta3-62715-05. Commit Hash: db02128e6e3c4bdfc93e6ec425ac9162b4d4fe80
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual Studio Code Debug Adapter Host Package   1.0
Interop layer for hosting Visual Studio Code debug adapters in Visual Studio

VisualStudio.Mac   1.0
Mac Extension for Visual Studio

Xamarin   4.9.0.753 (f0f46392f)
Visual Studio extension to enable development for Xamarin.iOS and Xamarin.Android.

Xamarin Designer   4.10.96 (a702a9ceb)
Visual Studio extension to enable Xamarin Designer tools in Visual Studio.

Xamarin.Android SDK   8.2.0.16 (HEAD/a78295902)
Xamarin.Android Reference Assemblies and MSBuild support.

Xamarin.iOS and Xamarin.Mac SDK   11.9.1.24 (f62de47)
Xamarin.iOS and Xamarin.Mac Reference Assemblies and MSBuild support.

**Comment by jalbertSyncroTech** [comment]
@stmax82 We're seeing this issue as well, along with other issues related to Android Resources.  I've reported both issues and have a work-around.

#1399 Visual Studio 15.6.1 regression Android.Util.AndroidRuntimeException Activity did not call through to super.onCreate()
- We also saw the "App works with rebuild, doesn't work with build" issue here - doing a rebuild after every change was the only way to avoid the issue.

#1544 Android.Runtime.ResourceIdManager.UpdateIdValues() called multiple times, adds 23 seconds to application startup
- Moving all of our resources out of the libraries and into the application, in addition to improving the startup time, also fixed the above issue.

**Comment by justintoth** [comment]
Same issue here, but only on release builds, where a resource that clearly exists can't be found (FindViewById returns null.)

**Comment by JonDouglas** [comment]
I'm not able to reproduce this issue with the given reproduction steps @stmax82. However with that said, I believe this issue is related around `bin/obj` state. Before I go further into that, here are my results:

1. Use Case 1
![try1](https://user-images.githubusercontent.com/1288848/39533842-54c0adae-4ded-11e8-911f-19aa8a7bbd95.png)

2. Use Case 2
![try2](https://user-images.githubusercontent.com/1288848/39533849-57010e4c-4ded-11e8-8d43-de47948cee96.png)

So unfortunately it looks like it's finding the `View` just fine in both cases on the latest 15.6 build. I even tried on 15.7 previews and it seems fine as well. 

However I have seen this issue in various forms in which majority of the issue comes down to a "bad cache" inside the `bin/obj` folders that potentially mismatches resources. Your symptoms of:

- Sometimes this line correctly returns a SurfaceView instance.
- Sometimes it returns null (and causes NullReferenceExceptions further down).
- Sometimes it throws an InvalidCastException because it tries to return an instance of something else.

All match what I've seen in various projects. Although clearing out `bin/obj` folders of your project will typically workaround this problem, this isn't a great solution. We are currently looking into solutions for this general problem.

What can really help assist us in figuring out this issue is to send us projects that are in the "bad cache" state with the existing `bin/obj` folders in tact given the intermittent behavior as described in the three bullets above. I am marking this issue as a bug given this is a known issue that we are actively looking for solutions to.

**Comment by stmax82** [comment]
@JonDouglas we tested this again on two different systems after updating Visual Studio to 15.7 and we were still able to reproduce the problem.

We also tested it with different phones - Nexus 5X, Samsung Galaxy S5 and in the android emulator.

It might take a bit, but I'll try to get you the bin/obj folders of the failing builds.

What else could cause this problem on our systems, while it seems to work on yours?

**Comment by JonDouglas** [comment]
@stmax82 I believe we made some findings recently that might improve this experience:

https://github.com/xamarin/xamarin-android/pull/1660

As for being a problem on your systems vs. mine, I would suspect it has to deal around cleans and rebuilds given that I only built and rebuilt the project a couple times and deployed it. I would suspect your environment has a few extra steps I may have missed. I'm hoping this helps out once we can get this in a proper release.
